src/postw90/berry.F90

```diff
@@ -1992,43 +2014,82 @@ subroutine berry_get_js_k(kpt, eig, del_alpha_eig, D_alpha_h, UU, js_k)
-      ! to get spin current, we need to multiply it by hbar/2,
-      ! also we need to divide it by hbar to recover the velocity
-      ! operator, these are done outside of this subroutine
-      js_k = 1.0_dp/2.0_dp*(B_k + conjg(transpose(B_k)))
+      if (index(shc_method, 'qiao') > 0) then !if Qiao
+        !=========== K_k ===========
+        ! < u_k | sigma_gamma | \partial_alpha u_k >, QZYZ18 Eq.(26)
+        ! QZYZ18 Eq.(37)
+        call pw90common_fourier_R_to_k_vec(kpt, SR_R(:, :, :, shc_gamma, :), OO_true=SR_w)
+        ! QZYZ18 Eq.(31)
+        SR_alpha_k = -cmplx_i*utility_rotate(SR_w(:, :, shc_alpha), UU, num_wann)
+        K_k = SR_alpha_k + matmul(S_k, D_alpha_h)
+
+        !=========== L_k ===========
+        ! < u_k | sigma_gamma.H | \partial_alpha u_k >, QZYZ18 Eq.(27)
+        ! QZYZ18 Eq.(38)
+        call pw90common_fourier_R_to_k_vec(kpt, SHR_R(:, :, :, shc_gamma, :), OO_true=SHR_w)
+        ! QZYZ18 Eq.(32)
+        SHR_alpha_k = -cmplx_i*utility_rotate(SHR_w(:, :, shc_alpha), UU, num_wann)
+        ! QZYZ18 Eq.(39)
+        call pw90common_fourier_R_to_k_vec(kpt, SH_R, OO_true=SH_w)
+        ! QZYZ18 Eq.(32)
+        SH_k = utility_rotate(SH_w(:, :, shc_gamma), UU, num_wann)
+        L_k = SHR_alpha_k + matmul(SH_k, D_alpha_h)
+
+        !=========== B_k ===========
+        ! < \psi_nk | sigma_gamma v_alpha | \psi_mk >, QZYZ18 Eq.(24)
+        B_k = cmplx_0
+        do i = 1, num_wann
+          eig_mat(i, :) = eig(:)
+          del_eig_mat(i, :) = del_alpha_eig(:)
+        end do
+        ! note * is not matmul
+        B_k = del_eig_mat*S_k + eig_mat*K_k - L_k
+
+        !=========== js_k ===========
+        ! QZYZ18 Eq.(23)
+        ! note the S in SR_R,SHR_R,SH_R of get_SHC_R is sigma,
+        ! to get spin current, we need to multiply it by hbar/2,
+        ! also we need to divide it by hbar to recover the velocity
+        ! operator, these are done outside of this subroutine
+        js_k = 1.0_dp/2.0_dp*(B_k + conjg(transpose(B_k)))
+
+      else !if Ryoo (PRB RPS19 Eq.(21))
+        do j = shc_alpha, shc_alpha
+          do i = shc_gamma, shc_gamma !RPS19 Eqs.(37)-(40)
+            call pw90common_fourier_R_to_k_new(kpt, SAA_R(:, :, :, i, j), OO=SAA(:, :, i, j))
+            call pw90common_fourier_R_to_k_new(kpt, SBB_R(:, :, :, i, j), OO=SBB(:, :, i, j))
+          end do
+        end do
+
+        call pw90common_fourier_R_to_k_new(kpt, HH_R, OO=HH, &
+                                           OO_dx=delHH(:, :, 1), &
+                                           OO_dy=delHH(:, :, 2), &
+                                           OO_dz=delHH(:, :, 3))
+
+        VV0(:, :) = utility_rotate(delHH_alpha(:, :), UU, num_wann)
+        !if(.true.) write(*,*) "kpt",kpt,"SA_K",real(SAA(1,:,1,3),dp)
+        do i = 1, 3
+          do j = 1, 3
+            SAA(:, :, i, j) = utility_rotate(SAA(:, :, i, j), UU, num_wann)
+            SBB(:, :, i, j) = utility_rotate(SBB(:, :, i, j), UU, num_wann)
```

Comment:
This part seems to be not consistent with the `call pw90common_fourier_R_to_k_new` part, where only `SBB(:, :, shc_gamma, shc_alpha)` is calculated.

Reply:
modified as you mentioned.